### DISCUSSION TITLE
How to receive and parse messages from UART in APP layer?

### DISCUSSION THREAD

**Discussion by Residualstress** [bitcraze/discussions#1105] [Q&A] [answered]
Hello,
I'm currently building the connection between Crazybolt and Raspberry pi. In my experiment, Raspberry Pi only sends the object detection result (i.e., the location of an object (55, 80)) to Crazybolt. And the Crazybolt will get the detection results and further process it. Now the work on Raspberry Pi and physical connection have been completed.

I indeed notice some suggestions on similar question, such as https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/development/uart_communication/. But this tutorial is for someone who wants to control crazyflie bu UART on Raspberry Pi. I also found a tutorial about Serial input basics on Arduino, https://forum.arduino.cc/t/serial-input-basics-updated/382007. But it doesn't work on APP layer.

For my problem, I don't need Raspberry Pi to control the crazyflie since its only task is to send coordinate information. And all the control commands will be executed on APP Layer by Crazybolt itself.

So, my question is how to receive and parse messages from UART in APP layer?  Actually, I have successfully implemented these examples: app_hello_world, app_push_demo. Could you please give me some advice on this?

Best regards,
Weijie

**Comment by knmcguire** [comment, marked as answer]
Hi!  Indeed I would advise [this tutorial](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/development/uart_communication/) that you already have given. It is for CRTP connection, so you can do anything with it that you want. In your case, I would recommend setting up a [logging parameter](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/userguides/logparam/) in the app layer and read it out in the [cflib logging retrieval](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/user-guides/python_api/#logging).  Then you'll only read out the logging from the bolt, but you can still control the bolt from the app layer alone. 

However, the top solution will block any other CRTP connections from other sources like your computer, as there can only be a 1-1 connection with crtp... So perhaps that is also another reason to go more low level.

As you have already seen in the tutorial explained before, that the abbrivation CPX comes up. That stands for [Crazyflie Packet exchange](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/cpx/) and that is actually a package protocol that wraps the messages in packages and makes sure it reads it out at the end. that is very flexible and can be used for simple messages as well. For instance, we use it for the [AIdeck video streaming example](https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/simple-examples/wifi-streamer/), where the package only contains the image information. This is an uart information sending example of the AIdeck with CPX, which perhaps something that you can look at for some inspiration: https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/other-examples/stm_gap8_cpx/. 

If you want to go one step lower (and since you only have 2 values that might be fine), then I thinkt that the uart simple example of the AIdeck still exists:
- https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/other-examples/uart-send-char/ (force compiling shouldn't be necessary anymore... that should be updated).
- I can't find the uart app layer equivalent anymore though... It might have disappeared. 

Anyway, it would be easier to use the first 2 option as there is already a message protocol to use instead of making it yourself, especially since you are sending 2 values and not 1.

**Comment by Pitro979** [comment]
Hello,
I would like to get a similar setup working using the following hardware:
- Luckfox Pico Mini B
- Crazyflie 2.1 (not BOLT)
 
The product page for the board can be found here: https://www.luckfox.com/Luckfox-Pico-Mini-B, and in the wiki  here: https://wiki.luckfox.com/Luckfox-Pico/Luckfox-Pico-UART.

With the goal of using the Luckfox board (onboard the drone since it's tiny) as the "client" for the drone, as if it was a PC connecting over Crazyradio, using serial communication instead of CR. 
We would use one of the expansion connectors on the CF to connect to a UART interface on the Luckfox, for instance pins 10 and 11 for UART3.
We have some doubts regarding power delivery to the board which can be either handled via USB, or the VBUS and ground lines on pins 1 and 2, which would need a 5V and ground rail respectively. 

In particular we have the following questions:
- Which is the best way to power the Luckfox board onboard the CF?
- Would the prototyping deck or expansion deck be needed to access relevant CF pins?
- What would be needed (as software) on the CF side to send commands and receive telemetry data between Luckfox and CF over UART?

**Reply by knmcguire** [reply]
Could you start a new discussion please? This thread has already been completed, so we prefer to handle support in a new one.
